### ISSUE TITLE
The `valid-jsdoc` rule and jsdoc comments for fields.

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#2938]
I tried to write a jsdoc comment for a field, but the comment seems to be bonded to a function in its right hand expression.

On the following code, I expected the comment `"Description for this.xs"` is bonded to `this.xs`, but actually the comment seems to be bonded to `x => x != null`.

```js
/**
 * Description for A.
 */
class A {
  /**
   * Description for constructor.
   * @param {object[]} xs - xs
   */
  constructor(xs) {
    /**
     * Description for this.xs;
     * @type {object[]}
     */
    this.xs = xs.filter(x => x != null);
  }
}
```

----

```
> eslint -v
v0.24.0
> eslint test.js --env es6 --rule "valid-jsdoc:[2,{requireReturn:false}]" --reset --no-eslintrc
```

## Actual

```
test.js
  10:4  error  Missing JSDoc for parameter 'x'  valid-jsdoc

✖ 1 problem (1 error, 0 warnings)
```

## Expected

No errors.

**Comment by ilyavolodin** [comment]
Could be related to either https://github.com/eslint/espree/issues/155 or https://github.com/eslint/espree/issues/136

**Comment by gyandeeps** [comment]
I dont think its related since the AST getting generated is correct. I verified it on my machine.
The issue is on this line https://github.com/eslint/eslint/blob/master/lib/eslint.js#L911
The second condition evaluates it true (actually it should be false) tho the object being compared are not equal. I think this check is bad as we are trying to compare to objects.

If we fix that line then the above scenario would work fine.

**Comment by gyandeeps** [comment]
UPDATE:
Once I removed the second check on line 911 (as mentioned above), all the unit test for all rules are passing including the scenario discussed in this issue.

Once issue gets accepted I will work on this.

**Comment by mysticatea** [comment]
Thank you a lot!

> I think obj1 !== obj2 would always return true because javascript object comparison is not correct/perfect.

In this case, I guess the comparison becomes `false` for IIFE, but becomes `true` for functions in the argument list :)
